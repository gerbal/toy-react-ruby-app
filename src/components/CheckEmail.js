import React from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import { StyleSheet, css } from "aphrodite";

import {
  updateName,
  updateCompanyDomain,
  fetchEmail,
  sendError
} from "../reducers/email-reducer";

const styles = StyleSheet.create({
  container: {
    display: "flex",
    flexFlow: "column",
    alignItems: "center",
    justifyContent: "center",
    fontFamily: '"Helvetica Neue", Helvetica, Verdana, Geneva, sans-serif'
  },
  block: {
    flex: 1,
    display: "flex",
    flexFlow: "row wrap",
    alignItems: "left",
    margin: "1rem",
    width: "100wv",
    maxWidth: "30rem"
  },
  inputLabel: {
    flex: "1 1 100%"
  },
  inputs: {
    flex: "1 0 100%",
    border: "1px solid #000000",
    padding: "0.5rem",
    margin: "0.5rem",
    width: "100vw",
    boxSizing: "border-box",
    maxWidth: "30rem",
    clear: "both"
  }
});

const InputWithLabel = ({ onChange, value, label }) => {
  return (
    <div className={css(styles.block)}>
      <label
        className={css(styles.inputLabel)}
        htmlFor={label}
      >
        {label}
      </label>
      <input
        className={css(styles.inputs)}
        id={label.replace(/\W/, '')}
        onChange={onChange}
        placeholder={label}
        value={value}
      />
    </div>
  );
};
InputWithLabel.propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.string
};

export class CheckEmail extends React.Component {
  constructor(props) {
    super(props);

    this.handleUpdateName = this.handleUpdateName.bind(this);
    this.handleUpdateDomain = this.handleUpdateDomain.bind(this);
    this.updateEmail = this.updateEmail.bind(this);
  }

  handleUpdateName = event => {
    const name = event.target.value;
    this.props.actions.updateName(name);
    this.updateEmail(name, this.props.email.domain);
  };

  handleUpdateDomain = event => {
    const domain = event.target.value;
    this.props.actions.updateCompanyDomain(domain);
    this.updateEmail(this.props.email.name, domain);
  };

  updateEmail = (name, domain) => {
    if (this.validDomain(domain)) {
      this.props.actions.fetchEmail(name, domain);
    } else if (!name) {
      this.props.actions.sendError("Missing Name");
    } else {
      this.props.actions.sendError("Invalid Domain");
    }
  };

  validDomain = domain => {
    return !!domain.match(/[\w]+(?:\.[\w]+)/);
  };

  render() {
    return (
      <div className={css(styles.container)}>
        <InputWithLabel
          label="Full Name"
          onChange={this.handleUpdateName}
          value={this.props.email.name}
        />
        <InputWithLabel
          label="Domain Name"
          onChange={this.handleUpdateDomain}
          value={this.props.email.domain}
        />

        {this.props.email.error && (
          <div className={css(styles.block)}>{this.props.email.error}</div>
        )}
        {this.props.email.email && (
          <div className={css(styles.block)}>{this.props.email.email}</div>
        )}
      </div>
    );
  }
}

CheckEmail.propTypes = {
  actions: PropTypes.object.isRequired,
  email: PropTypes.object.isRequired
};

function mapStateToProps(state, props) {
  return {
    email: state.email
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(
      { updateName, updateCompanyDomain, fetchEmail, sendError },
      dispatch
    )
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(CheckEmail);
