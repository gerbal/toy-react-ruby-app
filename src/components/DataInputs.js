import React from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import {
  updateName,
  updateCompanyDomain,
  fetchEmail
} from "../reducers/email-reducer";

class DataInputs extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      domain: ""
    };

    this.handleUpdateName = this.handleUpdateName.bind(this);
    this.handleUpdateDomain = this.handleUpdateDomain.bind(this);
    this.updateEmail = this.updateEmail.bind(this);
  }

  handleUpdateName = event => {
    this.setState({ name: event.target.value });
    this.props.actions.updateName(event.target.value);
    this.updateEmail();
  };

  handleUpdateDomain = event => {
    this.setState({ domain: event.target.value });
    this.props.actions.updateCompanyDomain(event.target.value);
    this.updateEmail();
  };

  updateEmail = () => {
    const name = this.state.name;
    const domain = this.state.domain;
    this.props.actions.fetchEmail(name, domain);
  };

  render() {
    console.log(this.props.email);
    return (
      <div>
        <input
          onChange={this.handleUpdateName}
          placeholder={"Name"}
          value={this.state.name}
        />
        <input
          onChange={this.handleUpdateDomain}
          placeholder={"Domain Name"}
          value={this.state.domain}
        />
        <span>{this.props.email.error && this.props.email.error}</span>
        <span>
          {this.props.email.email && this.props.email.email}
        </span>
      </div>
    );
  }
}

DataInputs.propTypes = {
  actions: PropTypes.object.isRequired,
  email: PropTypes.object.isRequired
};

function mapStateToProps(state, props) {
  return {
    email: state.email
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(
      { updateName, updateCompanyDomain, fetchEmail },
      dispatch
    )
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(DataInputs);
