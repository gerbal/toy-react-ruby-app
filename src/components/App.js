import React from "react";
import PropTypes from "prop-types";
import CheckEmail from "./CheckEmail";

const App = ({ name }) => {
  return (
    <div>
      <CheckEmail />
    </div>
  );
};

App.propTypes = {
  name: PropTypes.string
};

export default App;
