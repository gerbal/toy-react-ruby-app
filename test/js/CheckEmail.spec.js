import React from 'react';
import { mount } from 'enzyme';
import { StyleSheetTestUtils } from 'aphrodite';

import { CheckEmail } from '../../src/components/CheckEmail';

// aphrodite setup. If the project grows move to a testSetup.js or something
beforeEach(() => {
  StyleSheetTestUtils.suppressStyleInjection();
});
afterEach(() => {
  return new Promise(resolve => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
    return process.nextTick(resolve);
  });
});

describe('Component: CheckEmail', () => {
  it('should render the CheckEmail component', () => {
    const props = {
      email: {},
      actions: {}
    };
    const wrapper = mount(
      <CheckEmail {...props} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger call an error with no domain', () => {
    const updateName = jest.fn();
    const updateCompanyDomain = jest.fn();
    const fetchEmail = jest.fn();
    const sendError = jest.fn();

    const props = {
      email: {
        domain: '',
        name: ''
      },
      actions: { updateName, updateCompanyDomain, fetchEmail, sendError }
    };
    const wrapper = mount(
      <CheckEmail {...props} />
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.find('#FullName').simulate('change', { target: { value: 'name' } });
    expect (updateName).toHaveBeenCalledTimes(1);
    expect (sendError).toHaveBeenCalledTimes(1);
    expect (fetchEmail).not.toHaveBeenCalled();
  });

  it('should trigger call an error an invalid domain', () => {
    const updateName = jest.fn();
    const updateCompanyDomain = jest.fn();
    const fetchEmail = jest.fn();
    const sendError = jest.fn();

    const props = {
      email: {
        domain: '',
        name: ''
      },
      actions: { updateName, updateCompanyDomain, fetchEmail, sendError }
    };
    const wrapper = mount(
      <CheckEmail {...props} />
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.find('#FullName').simulate('change', { target: { value: 'name' } });
    expect (updateName).toHaveBeenCalledTimes(1);

    wrapper.find('#DomainName').simulate('change', { target: { value: 'name' } });
    expect (updateCompanyDomain).toHaveBeenCalledTimes(1);
    expect (sendError).toHaveBeenCalledTimes(2);
    expect (fetchEmail).not.toHaveBeenCalled();
  });

  it('should not trigger error callback if valid name and domain', () => {
    const updateName = jest.fn();
    const updateCompanyDomain = jest.fn();
    const fetchEmail = jest.fn();
    const sendError = jest.fn();

    const props = {
      email: {
        domain: '',
        name: ''
      },
      actions: { updateName, updateCompanyDomain, fetchEmail, sendError }
    };
    const wrapper = mount(
      <CheckEmail {...props} />
  );
    expect(wrapper).toMatchSnapshot();
    wrapper.find('#FullName').simulate('change', { target: { value: 'name' } });
    expect (updateName).toHaveBeenCalledTimes(1);

    wrapper.find('#DomainName').simulate('change', { target: { value: 'name.com' } });
    expect (updateCompanyDomain).toHaveBeenCalledTimes(1);
    expect (sendError).toHaveBeenCalledTimes(1);
    expect (fetchEmail).toHaveBeenCalledTimes(1);


    expect(wrapper).toMatchSnapshot();
  });
});
