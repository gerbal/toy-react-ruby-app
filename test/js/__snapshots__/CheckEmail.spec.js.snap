// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CheckEmail should not trigger error callback if valid name and domain 1`] = `
<CheckEmail
  actions={
    Object {
      "fetchEmail": [Function],
      "sendError": [Function],
      "updateCompanyDomain": [Function],
      "updateName": [Function],
    }
  }
  email={
    Object {
      "domain": "",
      "name": "",
    }
  }
>
  <div
    className="container_1ftli6h"
  >
    <InputWithLabel
      label="Full Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Full Name"
        >
          Full Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="FullName"
          onChange={[Function]}
          placeholder="Full Name"
          value=""
        />
      </div>
    </InputWithLabel>
    <InputWithLabel
      label="Domain Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Domain Name"
        >
          Domain Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="DomainName"
          onChange={[Function]}
          placeholder="Domain Name"
          value=""
        />
      </div>
    </InputWithLabel>
  </div>
</CheckEmail>
`;

exports[`Component: CheckEmail should not trigger error callback if valid name and domain 2`] = `
<CheckEmail
  actions={
    Object {
      "fetchEmail": [Function],
      "sendError": [Function],
      "updateCompanyDomain": [Function],
      "updateName": [Function],
    }
  }
  email={
    Object {
      "domain": "",
      "name": "",
    }
  }
>
  <div
    className="container_1ftli6h"
  >
    <InputWithLabel
      label="Full Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Full Name"
        >
          Full Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="FullName"
          onChange={[Function]}
          placeholder="Full Name"
          value=""
        />
      </div>
    </InputWithLabel>
    <InputWithLabel
      label="Domain Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Domain Name"
        >
          Domain Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="DomainName"
          onChange={[Function]}
          placeholder="Domain Name"
          value=""
        />
      </div>
    </InputWithLabel>
  </div>
</CheckEmail>
`;

exports[`Component: CheckEmail should render the CheckEmail component 1`] = `
<CheckEmail
  actions={Object {}}
  email={Object {}}
>
  <div
    className="container_1ftli6h"
  >
    <InputWithLabel
      label="Full Name"
      onChange={[Function]}
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Full Name"
        >
          Full Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="FullName"
          onChange={[Function]}
          placeholder="Full Name"
        />
      </div>
    </InputWithLabel>
    <InputWithLabel
      label="Domain Name"
      onChange={[Function]}
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Domain Name"
        >
          Domain Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="DomainName"
          onChange={[Function]}
          placeholder="Domain Name"
        />
      </div>
    </InputWithLabel>
  </div>
</CheckEmail>
`;

exports[`Component: CheckEmail should trigger call an error an invalid domain 1`] = `
<CheckEmail
  actions={
    Object {
      "fetchEmail": [Function],
      "sendError": [Function],
      "updateCompanyDomain": [Function],
      "updateName": [Function],
    }
  }
  email={
    Object {
      "domain": "",
      "name": "",
    }
  }
>
  <div
    className="container_1ftli6h"
  >
    <InputWithLabel
      label="Full Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Full Name"
        >
          Full Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="FullName"
          onChange={[Function]}
          placeholder="Full Name"
          value=""
        />
      </div>
    </InputWithLabel>
    <InputWithLabel
      label="Domain Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Domain Name"
        >
          Domain Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="DomainName"
          onChange={[Function]}
          placeholder="Domain Name"
          value=""
        />
      </div>
    </InputWithLabel>
  </div>
</CheckEmail>
`;

exports[`Component: CheckEmail should trigger call an error with no domain 1`] = `
<CheckEmail
  actions={
    Object {
      "fetchEmail": [Function],
      "sendError": [Function],
      "updateCompanyDomain": [Function],
      "updateName": [Function],
    }
  }
  email={
    Object {
      "domain": "",
      "name": "",
    }
  }
>
  <div
    className="container_1ftli6h"
  >
    <InputWithLabel
      label="Full Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Full Name"
        >
          Full Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="FullName"
          onChange={[Function]}
          placeholder="Full Name"
          value=""
        />
      </div>
    </InputWithLabel>
    <InputWithLabel
      label="Domain Name"
      onChange={[Function]}
      value=""
    >
      <div
        className="block_8osbox"
      >
        <label
          className="inputLabel_1jzahh6"
          htmlFor="Domain Name"
        >
          Domain Name
        </label>
        <input
          className="inputs_eh1wt9"
          id="DomainName"
          onChange={[Function]}
          placeholder="Domain Name"
          value=""
        />
      </div>
    </InputWithLabel>
  </div>
</CheckEmail>
`;
